{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f5ba6e26",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import seaborn as sns\n",
    "import bioframe as bf\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import seaborn as sns\n",
    "from tqdm import tqdm\n",
    "from scipy.stats import spearmanr, mannwhitneyu, fisher_exact\n",
    "import scipy.sparse as sp_sparse\n",
    "from scipy.special import expit as sigmoid\n",
    "from sklearn.metrics import roc_auc_score, precision_recall_curve, auc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "34b2f956",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>aragwas_score</th>\n",
       "      <th>maf</th>\n",
       "      <th>mac</th>\n",
       "      <th>study.name</th>\n",
       "      <th>study.phenotype.name</th>\n",
       "      <th>overBonferroni</th>\n",
       "      <th>overPermutation</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>16653911</td>\n",
       "      <td>43.734411</td>\n",
       "      <td>0.051896</td>\n",
       "      <td>26</td>\n",
       "      <td>('clim-gs10_raw_Full imputed genotype_amm',)</td>\n",
       "      <td>clim-gs10</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>16634821</td>\n",
       "      <td>35.267820</td>\n",
       "      <td>0.051896</td>\n",
       "      <td>26</td>\n",
       "      <td>('clim-gs10_raw_Full imputed genotype_amm',)</td>\n",
       "      <td>clim-gs10</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>16634858</td>\n",
       "      <td>35.267820</td>\n",
       "      <td>0.051896</td>\n",
       "      <td>26</td>\n",
       "      <td>('clim-gs10_raw_Full imputed genotype_amm',)</td>\n",
       "      <td>clim-gs10</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>16601028</td>\n",
       "      <td>34.882685</td>\n",
       "      <td>0.051896</td>\n",
       "      <td>26</td>\n",
       "      <td>('clim-gs10_raw_Full imputed genotype_amm',)</td>\n",
       "      <td>clim-gs10</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>16602311</td>\n",
       "      <td>32.632170</td>\n",
       "      <td>0.047904</td>\n",
       "      <td>24</td>\n",
       "      <td>('clim-gs10_raw_Full imputed genotype_amm',)</td>\n",
       "      <td>clim-gs10</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1225243</th>\n",
       "      <td>5</td>\n",
       "      <td>9055418</td>\n",
       "      <td>4.000051</td>\n",
       "      <td>0.041916</td>\n",
       "      <td>21</td>\n",
       "      <td>('clim-bio13_raw_Full imputed genotype_amm',)</td>\n",
       "      <td>clim-bio13</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1225244</th>\n",
       "      <td>5</td>\n",
       "      <td>9056844</td>\n",
       "      <td>4.000051</td>\n",
       "      <td>0.041916</td>\n",
       "      <td>21</td>\n",
       "      <td>('clim-bio13_raw_Full imputed genotype_amm',)</td>\n",
       "      <td>clim-bio13</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1225245</th>\n",
       "      <td>5</td>\n",
       "      <td>15646309</td>\n",
       "      <td>4.000035</td>\n",
       "      <td>0.013972</td>\n",
       "      <td>7</td>\n",
       "      <td>('clim-prec3_raw_Full imputed genotype_amm',)</td>\n",
       "      <td>clim-prec3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1225246</th>\n",
       "      <td>5</td>\n",
       "      <td>7726496</td>\n",
       "      <td>4.000002</td>\n",
       "      <td>0.437126</td>\n",
       "      <td>219</td>\n",
       "      <td>('clim-prec5_raw_Full imputed genotype_amm',)</td>\n",
       "      <td>clim-prec5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1225247</th>\n",
       "      <td>5</td>\n",
       "      <td>7726658</td>\n",
       "      <td>4.000002</td>\n",
       "      <td>0.437126</td>\n",
       "      <td>219</td>\n",
       "      <td>('clim-prec5_raw_Full imputed genotype_amm',)</td>\n",
       "      <td>clim-prec5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1225248 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom       pos  aragwas_score       maf  mac   \n",
       "0           1  16653911      43.734411  0.051896   26  \\\n",
       "1           1  16634821      35.267820  0.051896   26   \n",
       "2           1  16634858      35.267820  0.051896   26   \n",
       "3           1  16601028      34.882685  0.051896   26   \n",
       "4           1  16602311      32.632170  0.047904   24   \n",
       "...       ...       ...            ...       ...  ...   \n",
       "1225243     5   9055418       4.000051  0.041916   21   \n",
       "1225244     5   9056844       4.000051  0.041916   21   \n",
       "1225245     5  15646309       4.000035  0.013972    7   \n",
       "1225246     5   7726496       4.000002  0.437126  219   \n",
       "1225247     5   7726658       4.000002  0.437126  219   \n",
       "\n",
       "                                            study.name study.phenotype.name   \n",
       "0         ('clim-gs10_raw_Full imputed genotype_amm',)            clim-gs10  \\\n",
       "1         ('clim-gs10_raw_Full imputed genotype_amm',)            clim-gs10   \n",
       "2         ('clim-gs10_raw_Full imputed genotype_amm',)            clim-gs10   \n",
       "3         ('clim-gs10_raw_Full imputed genotype_amm',)            clim-gs10   \n",
       "4         ('clim-gs10_raw_Full imputed genotype_amm',)            clim-gs10   \n",
       "...                                                ...                  ...   \n",
       "1225243  ('clim-bio13_raw_Full imputed genotype_amm',)           clim-bio13   \n",
       "1225244  ('clim-bio13_raw_Full imputed genotype_amm',)           clim-bio13   \n",
       "1225245  ('clim-prec3_raw_Full imputed genotype_amm',)           clim-prec3   \n",
       "1225246  ('clim-prec5_raw_Full imputed genotype_amm',)           clim-prec5   \n",
       "1225247  ('clim-prec5_raw_Full imputed genotype_amm',)           clim-prec5   \n",
       "\n",
       "         overBonferroni  overPermutation ref alt  \n",
       "0                  True             True   A   C  \n",
       "1                  True             True   G   A  \n",
       "2                  True             True   G   A  \n",
       "3                  True             True   T   G  \n",
       "4                  True             True   C   T  \n",
       "...                 ...              ...  ..  ..  \n",
       "1225243           False            False   C   T  \n",
       "1225244           False            False   C   T  \n",
       "1225245           False            False   C   T  \n",
       "1225246           False            False   C   A  \n",
       "1225247           False            False   C   G  \n",
       "\n",
       "[1225248 rows x 11 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aragwas = pd.concat([\n",
    "    pd.read_csv(\n",
    "        f\"output/aragwas/api_chr{chrom}.csv.gz\",\n",
    "        usecols=[\"snp.chr\", \"snp.position\", \"snp.ref\", \"snp.alt\", \"score\", \"maf\", \"mac\",\n",
    "                 \"study.name\", \"study.phenotype.name\", \"overBonferroni\", \"overPermutation\"],\n",
    "    ).rename(columns={\n",
    "        \"snp.chr\": \"chrom\",\n",
    "        \"snp.position\": \"pos\",\n",
    "        \"snp.ref\": \"ref\",\n",
    "        \"snp.alt\": \"alt\",\n",
    "        \"score\": \"aragwas_score\",\n",
    "    }) for chrom in [\"1\", \"2\", \"3\", \"4\", \"5\"]],\n",
    "    #axis=1,\n",
    "    ignore_index=True\n",
    ")\n",
    "aragwas.chrom = aragwas.chrom.str.replace(\"chr\", \"\")\n",
    "aragwas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1b7e1803",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_aragwas = aragwas.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8d2a52d",
   "metadata": {},
   "outputs": [],
   "source": [
    "aragwas = aragwas.groupby([\"chrom\", \"pos\", \"ref\", \"alt\"]).agg({\n",
    "    \"aragwas_score\": \"max\", \"overBonferroni\": \"any\", \"overPermutation\": \"any\",\n",
    "    \"study.phenotype.name\": \"unique\",\n",
    "}).reset_index()\n",
    "aragwas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b50d948a",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = Path(f\"output/variants/all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abc301dd-4cd4-4254-bbf6-a324474c8150",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sgkit has square values, and upper triangular\n",
    "# vs. others\n",
    "\n",
    "LD_path = \"output/aragwas/LD_100000_0.1.npz\"\n",
    "\n",
    "\n",
    "LD = sp_sparse.load_npz(LD_path)\n",
    "LD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d0d8171-7c40-400e-8c79-0cfdbb7d8150",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(5,5))\n",
    "plt.imshow(LD[:1000, :1000].toarray())\n",
    "plt.colorbar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91816e31-6e0c-4bdc-b34d-026803af2194",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_symmetric(A):\n",
    "    return A + A.T - sp_sparse.diags(A.diagonal())\n",
    "\n",
    "def sparsify(A, threshold):\n",
    "    B = A.copy()\n",
    "    B.data[B.data < threshold] = 0\n",
    "    B.eliminate_zeros()\n",
    "    return B\n",
    "\n",
    "\n",
    "def square(A):\n",
    "    B = A.copy()\n",
    "    B.data = np.square(B.data)\n",
    "    return B\n",
    "\n",
    "# no need to make symmetric for LD vs. distance\n",
    "\n",
    "\n",
    "LD = make_symmetric(LD)  # for the ones from sgkit\n",
    "#LD = square(LD)\n",
    "LD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7163a53b-55a0-431b-9033-75f8f54dcf3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(5,5))\n",
    "plt.imshow(LD[:1000, :1000].toarray())\n",
    "plt.colorbar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "439c518a-f543-4c5d-b313-e99928e3b5ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert np.allclose(LD.diagonal(), 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13ae9655",
   "metadata": {},
   "outputs": [],
   "source": [
    "# models with scores defined on all variants\n",
    "models = [\n",
    "    #\"ConvNet_batch2048_weight1.0_secondpart_30k\",\n",
    "    \"ConvNet_batch2048_secondpart_30k\",\n",
    "    #\"ConvNet_batch2048_weight0_secondpart_30k\",\n",
    "    #\"vep/scratch/users/gbenegas/checkpoints/GPN_Arabidopsis_multispecies/ConvNet_ss_12k/checkpoint-12000\",\n",
    "]\n",
    "\n",
    "# models with scores defined on aligned regions\n",
    "models_subset = [\n",
    "    #\"ConvNet_batch2048_weight1.0_secondpart_30k\",\n",
    "    \"ConvNet_batch2048_secondpart_30k\",\n",
    "    #\"ConvNet_batch2048_weight0_secondpart_30k\",\n",
    "    #\"vep/scratch/users/gbenegas/checkpoints/GPN_Arabidopsis_multispecies/ConvNet_ss_12k/checkpoint-12000\",\n",
    "    \"phyloP\",\n",
    "    \"phastCons\",\n",
    "]\n",
    "model_renaming = {\n",
    "    \"ConvNet_batch2048_secondpart_30k\": \"GPN\",\n",
    "    #\"ConvNet_batch2048_weight1.0_secondpart_30k\": \"repeat-weight-1\",\n",
    "    #\"ConvNet_batch2048_secondpart_30k\": \"repeat-weight-0.1\",\n",
    "    #\"ConvNet_batch2048_weight0_secondpart_30k\": \"repeat-weight-0\",\n",
    "    #\"ConvNet_batch2048_secondpart_30k\": \"8 species\",\n",
    "    #\"vep/scratch/users/gbenegas/checkpoints/GPN_Arabidopsis_multispecies/ConvNet_ss_12k/checkpoint-12000\": \"1 species\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b42d4e1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants = pd.read_parquet(d / \"variants.parquet\")\n",
    "\n",
    "for m in models_subset:\n",
    "    model_path = d / f\"{m}.parquet\"\n",
    "    model_name = model_renaming.get(m, m)\n",
    "    variants[model_name] = pd.read_parquet(model_path)[\"score\"].values\n",
    "#variants[\"1\"] = -np.ones_like(variants.phyloP)\n",
    "#models.append(\"1\")\n",
    "#models_subset.append(\"1\")\n",
    "variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78ab969c-cbf0-4ae7-88fd-91cca1cf53a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "models = [model_renaming.get(m, m) for m in models]\n",
    "models_subset = [model_renaming.get(m, m) for m in models_subset]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96b10222",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants = variants.merge(aragwas, how=\"inner\", on=[\"chrom\", \"pos\", \"ref\", \"alt\"], sort=True)\n",
    "variants[\"% SNPs are GWAS hits\"] = 100 * variants.overPermutation\n",
    "variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff6a87ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "repeats = pd.read_csv(\"input/repeats.bed.gz\", sep=\"\\t\").rename(columns=dict(genoName=\"chrom\", genoStart=\"start\", genoEnd=\"end\"))\n",
    "repeats.chrom = repeats.chrom.str.replace(\"Chr\", \"\")\n",
    "variants[\"start\"] = variants.pos\n",
    "variants[\"end\"] = variants.start + 1\n",
    "variants = bf.coverage(variants, repeats).rename(columns={\"coverage\": \"repeat\"})\n",
    "variants.repeat = variants.repeat.astype(bool)\n",
    "variants.repeat.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57f91624",
   "metadata": {},
   "outputs": [],
   "source": [
    "# In case conservation scores are undefined due to lack of alignments\n",
    "# There are a lot. Should consider doing the plot separately without phyloP\n",
    "# and phastCons to have full variants? At least for correlation matrix multiplication\n",
    "# The missingness pattern is very biologically relevant,\n",
    "# lack of alignment might mean new Arabidopsis-specific region, for example\n",
    "# GPN OR is actually much higher when including unalignable region\n",
    "variants_subset = variants.dropna().copy()\n",
    "variants_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "615e94ef-6e4e-47dd-bfef-008b9a65e8f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def LD_profile(LD, score):\n",
    "    return LD @ score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e4d3170-a89c-446d-abc3-dea8e07e942d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# very small number of ties (less than 2%), so just taking an arbitrary subset is ok\n",
    "(variants_subset[models_subset] <= variants_subset[models_subset].quantile(0.01)).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a812c563-b3eb-4591-b342-7aad16caaacc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def transform(score):    \n",
    "    # option 1:\n",
    "    #return score - score.max()\n",
    "\n",
    "    # option 2:\n",
    "    #return -score.abs()\n",
    "    \n",
    "    # option 3\n",
    "    res = np.zeros_like(score)\n",
    "    q = 0.01\n",
    "    bottom_idx = np.argsort(score)[:int(len(score)*q)]\n",
    "    res[bottom_idx] = -1.0\n",
    "    return res\n",
    "\n",
    "for m in models:\n",
    "    variants[m] = transform(variants[m])\n",
    "    \n",
    "for m in models_subset:\n",
    "    variants_subset[m] = transform(variants_subset[m])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f7accf1-b01b-4612-8d44-0b0637855a53",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "(variants[models] == -1).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddfd45c7-287f-4f51-911b-d4bdf0f0aa26",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "(variants_subset[models_subset] == -1).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72e1bc5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for model in models.copy():\n",
    "    name = fr\"{model}$\\times$LD\" if model != \"1\" else \"LD Score\"\n",
    "    variants[name] = LD_profile(LD, variants[model])\n",
    "    models.append(name)\n",
    "variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f1270eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "phylop_idx = np.where(~variants.phyloP.isna().values)[0]\n",
    "for model in models_subset.copy():\n",
    "    label = fr\"{model}$\\times$LD\"\n",
    "    variants_subset[label] = LD_profile(LD[phylop_idx][:, phylop_idx], variants_subset[model])\n",
    "    models_subset.append(label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5718522c-25ec-4a7b-8108-b161c823502b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cannot flip sign of GPN scores because phyloP and phastCons used the ref and cannot be recomputed\n",
    "variants = variants.query(\"AF < 0.5\")\n",
    "variants_subset = variants_subset.query(\"AF < 0.5\")\n",
    "variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a02fce24-365b-4ddf-a506-3593504cd293",
   "metadata": {},
   "outputs": [],
   "source": [
    "#variants[\"bin\"] = pd.qcut(variants[r\"GPN$\\times$LD\"], 100, labels=False)\n",
    "variants[\"bin\"] = pd.cut(variants[r\"GPN$\\times$LD\"], 10).apply(lambda x: x.mid)\n",
    "\n",
    "\n",
    "\n",
    "plt.figure(figsize=(2,2))\n",
    "g = sns.lineplot(\n",
    "    data=variants,\n",
    "    x=\"bin\",\n",
    "    y=\"% SNPs are GWAS hits\",\n",
    "    errorbar=None,\n",
    ")\n",
    "plt.xlabel(r\"GPN$\\times$LD percentile\")\n",
    "#plt.xlabel(r\"GPN$\\times$LD\")\n",
    "sns.despine();\n",
    "#plt.savefig(\"output/plots/ld_profile.svg\", bbox_inches='tight');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68009e3f-44ae-49d3-b5fa-faf79cbf7299",
   "metadata": {},
   "outputs": [],
   "source": [
    "ld_models = [m for m in models if \"LD\" in m]\n",
    "ld_models_subset = [m for m in models_subset if \"LD\" in m]\n",
    "ld_models, ld_models_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21e5cba2-84bd-46d3-9573-aa407492ad51",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_odds_ratio(df, threshold_quantile, model_name):\n",
    "    df = df.copy()\n",
    "    threshold = df[model_name].quantile(threshold_quantile)\n",
    "    df[\"ModelScoreTier\"] = \"NA\"\n",
    "    df.loc[df[model_name] <= threshold, \"ModelScoreTier\"] = \"Low\" \n",
    "    #df.loc[df[model_name] > df[model_name].quantile(0.1), \"ModelScoreTier\"] = \"High\"\n",
    "    df.loc[df[model_name] > threshold, \"ModelScoreTier\"] = \"High\"\n",
    "    df = df[df.ModelScoreTier.isin([\"Low\", \"High\"])]\n",
    "    \n",
    "    df.ModelScoreTier = df.ModelScoreTier.astype('category')\n",
    "    group_counts = df.groupby([\"overPermutation\", \"ModelScoreTier\"]).size()\n",
    "    group_counts2 = group_counts.copy()\n",
    "    #print(group_counts2)\n",
    "    group_counts = group_counts.loc[pd.MultiIndex.from_tuples(\n",
    "        [[False, 'High'],\n",
    "         [False, 'Low'],\n",
    "         [True, 'High'],\n",
    "         [True, 'Low']]\n",
    "    )].values.reshape((2,2))\n",
    "    min_counts = group_counts.min()\n",
    "    odds_ratio, p_value = fisher_exact(group_counts, alternative='greater')\n",
    "    if np.isinf(odds_ratio):\n",
    "        print(odds_ratio, p_value, group_counts2)\n",
    "    return odds_ratio, p_value, min_counts\n",
    "\n",
    "#threshold_quantiles = np.linspace(1e-2, 1e-1, 4)\n",
    "threshold_quantiles = np.logspace(-3, -1, 10)\n",
    "\n",
    "\n",
    "rows = []\n",
    "rows_subset = []\n",
    "\n",
    "for tq in threshold_quantiles:\n",
    "    for m in ld_models:\n",
    "        print(m, tq)\n",
    "        OR, p_value, min_counts = compute_odds_ratio(variants, tq, m)\n",
    "        rows.append([m, tq, OR, p_value, min_counts])\n",
    "    for m in ld_models_subset:\n",
    "        print(m, tq)\n",
    "        OR, p_value, min_counts = compute_odds_ratio(variants_subset, tq, m)\n",
    "        rows_subset.append([m, tq, OR, p_value, min_counts])\n",
    "results = pd.DataFrame(rows, columns=[\"Model\", \"Threshold quantile\", \"Odds ratio\", \"p_value\", \"min_counts\",])\n",
    "results[\"significant\"] = results.p_value <= 0.05\n",
    "results_subset = pd.DataFrame(rows_subset, columns=[\"Model\", \"Threshold quantile\", \"Odds ratio\", \"p_value\", \"min_counts\",])\n",
    "results_subset[\"significant\"] = results_subset.p_value <= 0.05\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32faad2c-4e5e-42bd-97b1-88c1cad453ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5c8a188-1d5b-4cc3-bbb6-79f7a8119375",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(3,3))\n",
    "g = sns.lineplot(\n",
    "    data=results_subset[results_subset.significant],\n",
    "    x=\"Threshold quantile\",\n",
    "    y=\"Odds ratio\",\n",
    "    hue=\"Model\",\n",
    "    hue_order=[r\"GPN$\\times$LD\", r\"phyloP$\\times$LD\", r\"phastCons$\\times$LD\"],#, \"LD Score\"],\n",
    "    #style=\"significant\",\n",
    "    #style_order=[True, False],\n",
    "    marker=\"o\",\n",
    ")\n",
    "g.axhline(1, ls='--', color=\"grey\")\n",
    "sns.despine();\n",
    "g.set(xscale='log');\n",
    "#plt.savefig(\"output/plots/vep_gwas_line.pdf\", bbox_inches='tight');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a0f8967-cb58-4c92-8d1f-e1b123873234",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(2,2))\n",
    "\n",
    "g = sns.barplot(\n",
    "    data=results_subset[(results_subset.p_value<=0.05)&(results_subset[\"Threshold quantile\"]==1e-3)],#1e-2)],\n",
    "    x=\"Model\",\n",
    "    y=\"Odds ratio\",\n",
    "    color=\"C0\",\n",
    ")\n",
    "g.set_xticklabels(g.get_xticklabels(), rotation=45, ha='right')\n",
    "sns.despine();\n",
    "g.set(ylim=(1));\n",
    "plt.xlabel(\"\")\n",
    "#plt.savefig(\"output/plots/vep_gwas_bar.svg\", bbox_inches='tight');\n",
    "#plt.savefig(\"output/plots/vep_gwas_bar_repeat_weight.svg\", bbox_inches='tight');\n",
    "#plt.savefig(\"output/plots/vep_gwas_bar_n_species.svg\", bbox_inches='tight');"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "975531e3",
   "metadata": {},
   "source": [
    "## Interpreting specific associations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c24e8c13-47d4-48bc-a94d-6c1772e6c007",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_phenotypes = full_aragwas[\"study.phenotype.name\"].value_counts().head(10).index.values\n",
    "full_aragwas.groupby([\"study.phenotype.name\", \"overPermutation\"]).size().loc[top_phenotypes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "852ed957-6969-4508-bdcd-7ccd229885fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_phenotype = \"clim-tmax1\"\n",
    "target_gwas = full_aragwas[full_aragwas[\"study.phenotype.name\"]==target_phenotype]\n",
    "target_gwas.overPermutation.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77c5b733-a627-471d-8323-3cc7e495e64f",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants[\"original_index\"] = variants.index\n",
    "variants_gwas = variants.drop(\n",
    "    columns=[\"aragwas_score\", \"overBonferroni\", \"overPermutation\",\"study.phenotype.name\",]\n",
    ").merge(target_gwas, how=\"inner\", on=[\"chrom\", \"pos\", \"ref\", \"alt\"]).set_index(\"original_index\")\n",
    "variants_gwas.overPermutation.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b17fbf37",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants_gwas[\"GWAS hit\"] = variants_gwas.overPermutation.replace({True: \"Yes\", False: \"No\"})\n",
    "variants_gwas[\"Chromosome position\"] = variants_gwas.pos\n",
    "\n",
    "def plot_variants_profile(vs, scores, save_path=None):\n",
    "    #plt.figure(figsize=(4,3))\n",
    "    ld_with_target = LD[vs.sort_values(scores[0]).index.values[0]][:, vs.index.values].toarray().ravel().round(decimals=1)\n",
    "    print(ld_with_target)\n",
    "    ld_col = r\"\"\"$r^2$ w/ lowest\n",
    "GPN score\"\"\"\n",
    "    vs[ld_col] = ld_with_target\n",
    "    \n",
    "    vs = pd.melt(\n",
    "        vs, [\"chrom\", \"Chromosome position\", \"ref\", \"alt\", \"GWAS hit\", ld_col], scores, \"Model\", \"Score\"\n",
    "    )\n",
    "    sns.set_style(\"darkgrid\") # dark  # to show the high ld with contrast\n",
    "    g = sns.relplot(\n",
    "        data=vs, x=\"Chromosome position\", y=\"Score\",\n",
    "        row=\"Model\",\n",
    "        #hue=\"GWAS hit\", hue_order=[True, False], palette=\"Dark2\",\n",
    "        style=\"GWAS hit\", style_order=[\"Yes\", \"No\"], markers=[\"o\", \"X\"], #markers=[\"X\", \".\"],\n",
    "        hue=ld_col, palette=\"viridis\",\n",
    "        #hue=\"GWAS hit\", hue_order=[\"Yes\", \"No\"], palette={\"Yes\": \"g\", \"No\": \"r\"},\n",
    "        kind=\"scatter\", facet_kws={'sharey': False, 'sharex': True},\n",
    "        height=2,\n",
    "        aspect=1.6,\n",
    "        s=100,\n",
    "    )\n",
    "    g.map(lambda **kwargs: plt.axhline(y=0, color='grey', linestyle='--'))\n",
    "    g.set_titles(col_template=\"\", row_template=\"\")\n",
    "    for i, score in enumerate(scores):\n",
    "        g.axes[i,0].set_ylabel(score)\n",
    "    plt.subplots_adjust(hspace=0.1)\n",
    "    #sns.move_legend(\n",
    "    #    g, \"lower center\",\n",
    "    #    bbox_to_anchor=(0.4, 0.9),\n",
    "    #    ncol=2,\n",
    "    #)\n",
    "    \n",
    "    #g.axhline(0, ls='--', color=\"grey\")  # would need to map?\n",
    "    #sns.move_legend(g, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "    if save_path is not None:\n",
    "        plt.savefig(save_path, bbox_inches=\"tight\")\n",
    "\n",
    "    \n",
    "def plot_variant_profile(v, score, context=100, **kwargs):\n",
    "    print(v.pos, context)\n",
    "    vs = variants_gwas[\n",
    "        (variants_gwas.pos >= v.pos-context) & (variants_gwas.pos <= v.pos+context)\n",
    "    ]\n",
    "    print(len(vs))\n",
    "    plot_variants_profile(vs, score, **kwargs)\n",
    "    \n",
    "\n",
    "def plot_ld_nearby(v, context=100):\n",
    "    vs = variants_gwas[\n",
    "        (variants_gwas.pos >= v.pos-context) & (variants_gwas.pos <= v.pos+context)\n",
    "    ]\n",
    "    plt.figure(figsize=(10, 10))\n",
    "    sns.heatmap(\n",
    "        pd.DataFrame(\n",
    "            LD[vs.index.values][:, vs.index.values].todense(),\n",
    "            index=vs.overPermutation, columns=vs.overPermutation\n",
    "        ),\n",
    "        square=True, xticklabels=True, yticklabels=True\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ffd10ce-891b-4b07-99b3-b80e6349c3ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def top_LD_with_variant(v, k=3):\n",
    "    row = LD.getrow(v.name)\n",
    "    values = row.data\n",
    "    indices = row.indices\n",
    "    top_indices = indices[np.argsort(-values)[:k]]\n",
    "    return variants_gwas[variants_gwas.index.isin(top_indices)]\n",
    "\n",
    "def LD_max_dist_with_variant(v, threshold=0.7):\n",
    "    print(LD[v.name,v.name])\n",
    "    row = LD.getrow(v.name)\n",
    "    values = row.data\n",
    "    indices = row.indices\n",
    "    top_indices = indices[values>threshold]\n",
    "    vs = variants_gwas[variants_gwas.index.isin(top_indices)]\n",
    "    return (v.pos-vs.pos).abs().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cca3a01",
   "metadata": {},
   "outputs": [],
   "source": [
    "# could also specify that there's no long-range LD (using a threshold)\n",
    "# there's a lot of long-range LD, to bear in mind\n",
    "# maybe filter the aragwas first for this\n",
    "# would maybe have to re-do the GWAS analysis to avoid long-range LD? (maybe using a subset of phenotypes)\n",
    "# by using PC's?\n",
    "# or maybe do the filtering of LD matrix based on a radius myself?\n",
    "# unclear what causes LD, to be honest.\n",
    "# or maybe just don't show the fine-mapping... if there's long range LD\n",
    "# the rest of the analysis is still good\n",
    "\n",
    "# climtmax-1 interesting examples:\n",
    "# chrom           5\n",
    "# pos      10360998\n",
    "# ref             T\n",
    "# alt             A\n",
    "# Name: 432170, dtype: object\n",
    "\n",
    "for _, v in variants_gwas.query('overPermutation').sort_values(\"GPN\").head(30).iterrows():\n",
    "    #print(v)\n",
    "    #max_abs_dist = (v.pos - top_LD_with_variant(v, k=2).pos).abs().max()\n",
    "    #print(max_abs_dist)\n",
    "    if v.pos != 24505063: continue\n",
    "    max_abs_dist = LD_max_dist_with_variant(v, 0.8)\n",
    "    if np.isnan(max_abs_dist):\n",
    "        print(v.name)\n",
    "        raise Exception(\"debug\")\n",
    "    if max_abs_dist == 0 or max_abs_dist > 10_000:\n",
    "        print(\"skipping\")\n",
    "        continue\n",
    "    print(v)\n",
    "    print(max_abs_dist)\n",
    "    plot_variant_profile(v, [\"GPN\", r\"GPN$\\times$LD\"], context=max_abs_dist+10)#, save_path=\"output/plots/gwas_example.svg\")\n",
    "    plt.show()\n",
    "    plot_ld_nearby(v, context=max_abs_dist+10)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84d0283a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "chrom                             5\n",
    "pos                        11531640\n",
    "ref                               T\n",
    "alt                               A\n",
    "AC                               12\n",
    "AN                             1686\n",
    "consequence      intergenic_variant\n",
    "AF                         0.007117\n",
    "GPN                      -10.289432\n",
    "phyloP                     0.472434\n",
    "phastCons                 -0.000185\n",
    "aragwas_score              7.353801\n",
    "aragwas_hit                    True\n",
    "bin_GPN                           0\n",
    "bin_phyloP                       37\n",
    "bin_phastCons                    44\n",
    "\"\"\";\n",
    "\n",
    "# inside repeat but close to small conserved region. will skip for now.\n",
    "# could compare to reference repeat.\n",
    "# could check if it's qtl in another database. issue is it depends on RNA-seq being\n",
    "# right tissue, developmental stage, stress, etc.\n",
    "# some nearby genes: AT5G31412, AT5G30360"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "320de507",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "chrom                             5\n",
    "pos                        11535787\n",
    "ref                               A\n",
    "alt                               T\n",
    "AC                               12\n",
    "AN                             1674\n",
    "consequence      intergenic_variant\n",
    "AF                         0.007168\n",
    "GPN                       -9.268337\n",
    "phyloP                    -0.315622\n",
    "phastCons                 -0.007427\n",
    "aragwas_score              11.85641\n",
    "aragwas_hit                    True\n",
    "bin_GPN                           0\n",
    "bin_phyloP                       17\n",
    "bin_phastCons                    28\n",
    "\"\"\";\n",
    "\n",
    "# not in a repeat. only aligned with the 2 other arabidopsis species\n",
    "# could try to understand if there's a motif there. also check the specific trait,\n",
    "# see if it's related to the nearby gene\n",
    "# motif here: seems like it's a strengthening of a binding site  (inside core motif)\n",
    "# http://planttfdb.gao-lab.org/tf.php?sp=Ath&did=AT3G46640.1#bind_motif\n",
    "# but if it's a strengthening then way would it receive a negative, instead of positive score\n",
    "# maybe don't showcase this\n",
    "# not a fine-mapping example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b8fb8f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "vs = variants[\n",
    "    (variants.pos >= v.pos-1000) & (variants.pos <= v.pos+1000)\n",
    "]\n",
    "plt.figure(figsize=(10, 10))\n",
    "sns.heatmap(\n",
    "    pd.DataFrame(\n",
    "        LD[vs.index.values][:, vs.index.values].todense(),\n",
    "        index=vs.overPermutation, columns=vs.overPermutation\n",
    "    ),\n",
    "    square=True, xticklabels=True, yticklabels=True\n",
    ")\n",
    "\"\"\";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85be8af9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "chrom                             5\n",
    "pos                        12172178\n",
    "ref                               G\n",
    "alt                               C\n",
    "AC                               10\n",
    "AN                             2222\n",
    "consequence      intergenic_variant\n",
    "AF                           0.0045\n",
    "GPN                       -9.151054\n",
    "phyloP                    -0.281095\n",
    "phastCons                 -0.061246\n",
    "aragwas_score              11.85641\n",
    "aragwas_hit                    True\n",
    "bin_GPN                           0\n",
    "bin_phyloP                       18\n",
    "bin_phastCons                    17\n",
    "\"\"\";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f202fbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "chrom                      5\n",
    "pos                  7543675\n",
    "ref                        C\n",
    "alt                        A\n",
    "AC                        24\n",
    "AN                      2198\n",
    "consequence      stop_gained\n",
    "AF                  0.010919\n",
    "GPN                -8.982161\n",
    "phyloP             -1.372914\n",
    "phastCons          -0.357317\n",
    "aragwas_score      14.915038\n",
    "aragwas_hit             True\n",
    "bin_GPN                    0\n",
    "bin_phyloP                 4\n",
    "bin_phastCons             10\n",
    "\"\"\";\n",
    "\n",
    "# gene: AT5G22690\n",
    "# pretty clear example, no other hits nearby:\n",
    "# https://aragwas.1001genomes.org/#/gene/AT5G22690"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "576b4895",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "chrom                             5\n",
    "pos                        12894850\n",
    "ref                               G\n",
    "alt                               C\n",
    "AC                                8\n",
    "AN                             2146\n",
    "consequence      intergenic_variant\n",
    "AF                         0.003728\n",
    "GPN                       -8.817844\n",
    "phyloP                    -1.393114\n",
    "phastCons                 -0.892624\n",
    "aragwas_score              11.85641\n",
    "aragwas_hit                    True\n",
    "bin_GPN                           0\n",
    "bin_phyloP                        4\n",
    "bin_phastCons                     5\n",
    "\"\"\";\n",
    "\n",
    "# lies in a conserved region, could check TFBS motif\n",
    "# e.g. here: http://plantregmap.gao-lab.org/binding_site_prediction.php\n",
    "# or could just download their database of non-conserved TFBS prediction\n",
    "# is predicted to bind this (make sure to put lowest threshold)\n",
    "# http://planttfdb.gao-lab.org/tf.php?sp=Ath&did=AT3G27785.1#bind_motif\n",
    "# be aware of reverse complement.\n",
    "# it matches part of the core motif\n",
    "# could also understand specific phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "282f5138",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "chrom                         5\n",
    "pos                    21174728\n",
    "ref                           A\n",
    "alt                           C\n",
    "AC                           58\n",
    "AN                         1854\n",
    "consequence      intron_variant\n",
    "AF                     0.031284\n",
    "GPN                   -7.741788\n",
    "phyloP                 -0.90847\n",
    "phastCons             -0.007009\n",
    "aragwas_score         16.127622\n",
    "aragwas_hit                True\n",
    "bin_GPN                       0\n",
    "bin_phyloP                    7\n",
    "bin_phastCons                29\n",
    "\"\"\";\n",
    "\n",
    "# interesting intron variant, could try check raw RNA-seq data coverage to see if a new splicing pattern\n",
    "# was created. although this is first intron, which also regulates expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc695a19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# overlaps a funtfbs, pretty weird conservation pattern though, would skip\n",
    "#v = variants.query('pos==18920189').iloc[0]\n",
    "#v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0165364-85a3-46c3-9216-c6f17500b106",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "len(variants_subset) / len(variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a47e98cf-18de-43ac-a788-17c15db856dd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gpn",
   "language": "python",
   "name": "gpn"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
